fn_query_tor_network/fn_query_tor_network/components/fn_tor.py

```diff
@@ -0,0 +1,55 @@
+# (c) Copyright IBM Corp. 2010, 2018. All Rights Reserved.
+# -*- coding: utf-8 -*-
+# pragma pylint: disable=unused-argument, no-self-use
+"""Function implementation"""
+
+import logging
+from resilient_circuits import ResilientComponent, function, handler, StatusMessage, FunctionResult, FunctionError
+import fn_query_tor_network.util.selftest as selftest
+import requests
+
+class FunctionComponent(ResilientComponent):
+    """Component that implements Resilient function 'fn_tor"""
+
+    def __init__(self, opts):
+        """constructor provides access to the configuration options"""
+        super(FunctionComponent, self).__init__(opts)
+        self.options = opts.get("fn_query_tor_network", {})
+        selftest.selftest_function(opts)
+
+    @handler("reload")
+    def _reload(self, event, opts):
+        """Configuration options have changed, save new values"""
+        self.options = opts.get("fn_query_tor_network", {})
+
+    @function("fn_tor")
+    def _fn_tor_function(self, event, *args, **kwargs):
+        """Function: This TOR function searches for the given IP Address or host names in TOR exit node Network by using RESTful API"""
+        try:
+            # Get the function parameters:
+            tor_search_data = kwargs.get("tor_search_data")  # text
+
+            log = logging.getLogger(__name__)
+            log.info("tor_search_data: %s", tor_search_data)
+
+            # PUT YOUR FUNCTION IMPLEMENTATION CODE HERE
+            yield StatusMessage("starting...")
+
+            __params_data = {'flag': self.options.get('flag'), 'fields': self.options.get('data_fields'),
+                             'search': tor_search_data}
+
+            __response_data = requests.get(self.options.get('base_url'), params=__params_data)
+
+            __response_data = __response_data.text
+            if __response_data.find(tor_search_data) != -1:
```

Comment:
I do want to change this logic to walk the json to confirm the ip address is found. How would that be done for domain name?

Reply:
1. The First Level of the search for given artifact will be done from the get API call, where we mentioned the data fields to search within and it only searched in those fields, for example if we mention data fileds = or_address,exit_address,hostname the api will search in these fields for given search data and return the JSON object as response to the get call.
2. if data found it will return below json where it will have data fields where it found the match.
  {
    "version": "7.0",
    "build_revision": "7dc8dae",
    "relays_published": "2018-12-21 08:00:00",
    "relays": [
        {
            "or_addresses": [
                "162.247.74.201:443"
            ],
            "exit_addresses": [
                "162.247.74.201"
            ]
        }
    ],
    "bridges_published": "2018-12-21 07:51:16",
    "bridges": []
}
3. if data is not found in the get Rest Api call it will return the below json object as response,
example api call
https://onionoo.torproject.org/details?flag=Exit&fields=exit_addresses,or_addresses,host_name&search=addresses
result : 
{
    "version": "7.0",
    "build_revision": "7dc8dae",
    "relays_published": "2018-12-21 08:00:00",
    "relays": [],
    "bridges_published": "2018-12-21 07:51:16",
    "bridges": []
}

so by looking at above both results it clear that it return only those fields where the data is getting matched if not it will not include those data fields on the response json object.

4. so i have directly searching artifact in response json object.if we make walk the json login also no use as above would be doing exactly same for us.
5. Here is the json walk logic to walk through the json data
#Walk Through JSON Object Code
            __response_data_json_obj = __response_data.json()
            __relays_data_list = __response_data_json_obj.get('relays')
            search_field_list = self.options.get('data_fields').split(',')
            __MATCH_FLAG = False
            if not __relays_data_list:
                log.info("No Object Matched...!")
                __MATCH_FLAG = False
            else:
                log.info("we found the data")
                for relay_data in __relays_data_list:
                    for field in search_field_list:
                        data = relay_data.get(field)
                        if data is not None:
                            if isinstance(data, list):
                                for element in data:
                                    if element.find(tor_search_data) != -1:
                                        log.info('Given Search Artifcat matched..!')
                                        __MATCH_FLAG = True
                            else:
                                if data.find(tor_search_data) != -1:
                                    log.info('Given Search Artifcat matched..!')
                                    __MATCH_FLAG = True
            if __MATCH_FLAG:
                results = {'status': 'success', 'value': True, 'data': __response_data.text}
            else:
                results = {'status': 'failed', 'value': False, 'data': __response_data.text}

            yield StatusMessage("done...")
            yield FunctionResult(results)